nixos/modules/programs/opengamepadui.nix

```diff
@@ -0,0 +1,268 @@
+      socket="''${tmpdir:+$tmpdir/startup.socket}"
+      stats="''${tmpdir:+$tmpdir/stats.pipe}"
+
+      # Fail early if we don't have a proper runtime directory setup
+      if [[ -z $tmpdir || -z ''${XDG_RUNTIME_DIR+x} ]]; then
+        echo >&2 "!! Failed to find run directory in which to create stats session sockets (is \$XDG_RUNTIME_DIR set?)"
+        exit 0
+      fi
+
+      export GAMESCOPE_STATS="$stats"
+      mkfifo -- "$stats"
+      mkfifo -- "$socket"
+
+      # Start gamescope compositor, log it's output and background it
+      echo gamescope ${lib.escapeShellArgs cfg.gamescopeSession.args}  -R $socket -T $stats >"$HOME"/.gamescope-cmd.log
+      gamescope ${lib.escapeShellArgs cfg.gamescopeSession.args} -R $socket -T $stats >"$HOME"/.gamescope-stdout.log 2>&1 &
+      gamescope_pid="$!"
+
+      if read -r -t 3 response_x_display response_wl_display <>"$socket"; then
+        export DISPLAY="$response_x_display"
+        export GAMESCOPE_WAYLAND_DISPLAY="$response_wl_display"
+        # We're done!
+      else
+        echo "gamescope failed"
+        kill -9 "$gamescope_pid"
+        wait -n "$gamescope_pid"
+        exit 1
+        # Systemd or Session manager will have to restart session
+      fi
+
+      # If we have mangoapp binary start it
+      if command -v mangoapp >/dev/null; then
+        (while true; do
+          sleep 1
+          mangoapp >"$HOME"/.mangoapp-stdout.log 2>&1
+        done) &
+      fi
+
+      # Start OpenGamepadUI
+      opengamepadui ${lib.escapeShellArgs cfg.args}
+
+      # When the client exits, kill gamescope nicely
+      kill $gamescope_pid
+    '';
+
+  gamescopeSessionFile =
+    (pkgs.writeTextDir "share/wayland-sessions/opengamepadui.desktop" ''
+      [Desktop Entry]
+      Name=opengamepadui
+      Comment=OpenGamepadUI Session
+      Exec=${opengamepadui-gamescope}/bin/opengamepadui-gamescope
+      Type=Application
+    '').overrideAttrs
+      (_: {
+        passthru.providedSessions = [ "opengamepadui" ];
+      });
+in
+{
+  options.programs.opengamepadui = {
+    enable = lib.mkEnableOption "opengamepadui";
+
+    args = lib.mkOption {
+      type = lib.types.listOf lib.types.str;
+      default = [ ];
+      description = ''
+        Arguments to be passed to OpenGamepadUI
+      '';
+    };
+
+    package = lib.mkPackageOption pkgs "OpenGamepadUI" {
+      default = [ "opengamepadui" ];
+    };
+
+    extraPackages = lib.mkOption {
+      type = lib.types.listOf lib.types.package;
+      default = [ ];
+      example = lib.literalExpression ''
+        with pkgs; [
+          gamescope
+        ]
+      '';
+      description = ''
+        Additional packages to add to the OpenGamepadUI environment.
+      '';
+    };
+
+    fontPackages = lib.mkOption {
+      type = lib.types.listOf lib.types.package;
+      # `fonts.packages` is a list of paths now, filter out which are not packages
```

Comment:
I'm confused as to why this comment is here. It looks like a backwards compatibility thing from another package...?

Reply:
Yes, I based the NixOS module on the Steam module:
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/steam.nix#L105-L116

Is this comment no longer accurate?

Reply:
The type is already `listOf package` which means that the check shouldn't be necessary